docs/smart_bits.md

```diff
@@ -0,0 +1,71 @@
+
+## Operators
+Arithmetic, bit-wise logical, comparison, shift, bit-wise reduction, and concat operators are defined for all smart types and intermediates. All operators are implemented as python-style infix operators (e.g. `a + b`), except for the following:
+* Concat: the function `concat` is a free, variadic function; e.g. `long = concat(x, y, z)`
+* Bit-wise reduce: bit-wise reductions are defined as a member function `reduce`, which takes the reduction operator as an argument; e.g. `short = x.reduce(operator.or_)`
+
+See the end of the document for a table summarizing all operators.
+
+## Inference
+We first define a *partial expression* to be any (potentially compound) expression with smart operands, without wiring, and we define a *complete expression* to be any wiring statement where the left hand side is a base smart type and the right hand side is a partial expression. Every complete expression is dynamically resolved using a scheme to infer the bit-width and signedness of each operand. This is inspired from the [Verilog standard](https://www.eg.bucknell.edu/~csci320/2016-fall/wp-content/uploads/2015/08/verilog-std-1364-2005.pdf) (see Sec. 5.2).
+
+In general, every partial expression appearing in a complete expression is either *self-determined* or *context-determined*. This distinguishes expressions which require the entirety of the complete expression in-order to do inference, or only the partial expression. Each operator may have further specialized inference rules.
+
+### Bit-width
+The rule of thumb for context-determined expressions is that all it's operands get extended to the **longest** operand appearing in the complete expression (including the LHS of the assignment). On the other hand, the scope for self-determined expressions is only the partial expression itself. Note that since a context-determined expression can appear within a self-determined expression, each self-determined expression must still be procssed by the inference step, except that the context is limited to just the expression itself, and in particular, the LHS of the assignment is excluded.
+
+The bit-width of any partial expression's result is determined by the specific operator (see table below). For example the result of addition has bit-width equal to the maximum of all of it's operand bit-widths, where as the result of comparison always has bit-width equal to 1.
+
+### Signedness
+The signedness of operands can both determine the semantics of operators, as well as how operands are extended. For example, signed-comparison has different semantics than unsigned-comparison, and operands can either be sign-extended or zero-extended (during bit-width inference).
```

Comment:
Given:
```Python
x = SmartBits[8, True]()
y = SmartBits[16, False]()
```
`x+y` is unsigned but is x sign extend or zero extended?

Reply:
Great question. From my reading of the standard it should be **zero-extended**:

> The following are the steps for evaluating an expression:
> * Determine the expression size based upon the standard rules of expression size determination.
> * Determine the sign of the expression using the rules outlined in 5.5.1.
> * Propagate the type and size of the expression (or self-determined sub expression) back down to the context-determined  operands  of  the  expression.  In  general,  any  context-determined  operand  of  an operator  shall  be  the  same  type  and  size  as  the  result  of  the  operator.  However,  there  are  two exceptions:
>   ...
> * When propagation reaches a simple operand as defined in 5.2 (a primary as defined in A.8.4), then that  operand  shall  be  converted  to  the  propagated  type  and  size.  If  the  operand  must  be  extended,then it shall be sign-extended only if the propagated type is signed.

I actually just checked and the code was **not** following this^. So I've now fixed it in a recent commit.

Reply:
Hmmm I suspect I like the old code better.  It feels more natural to me to have signedness of the sub expression determine the extension.  But I suspect other people would disagree (which is why I don't like these sort of coercion rules).   Anyway, is there some way to force extension so I could get this to work:
```Python
x = SmartBits[16, True](-1)
y = SmartBits[32, False](1)
assert x + y == 0
```

Reply:
Yes, that's the reason for the `signed()` and `unsigned()` coercion functions, so if you had

```python
x = SmartBits[16, True](-1)
y = SmartBits[32, False](1)
assert x + signed(y) == 0
```
it would work.

Reply:
But that's not quite the same.  As that would return a signed expression and do a signed version of the operator (not that it matters for plus).

What I was wondering if there was some way to do: `x.ext(y.size-x.size) + y`